From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 25 Dec 2022 19:31:09 +0100
Subject: [PATCH] Don't load chunks to spawn phantoms

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Gale - https://galemc.org

This patch is based on the following patch:
"Don't load chunks when spawning phantoms"
By: PureGero <puregero@gmail.com>
As part of: MultiPaper (https://github.com/MultiPaper/MultiPaper)
Licensed under: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

diff --git a/src/main/java/net/minecraft/world/level/levelgen/PhantomSpawner.java b/src/main/java/net/minecraft/world/level/levelgen/PhantomSpawner.java
index 021221da5d0315f6e371380a705ac6b3f6ac18d3..997f80d332b95bb011cbbc27c065b2811a2dddc7 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/PhantomSpawner.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/PhantomSpawner.java
@@ -70,7 +70,15 @@ public class PhantomSpawner implements CustomSpawner {
 
                                     if (randomsource.nextInt(j) >= world.paperConfig().entities.behavior.playerInsomniaStartTicks) { // Paper - Ability to control player's insomnia and phantoms
                                         BlockPos blockposition1 = blockposition.above(20 + randomsource.nextInt(15)).east(-10 + randomsource.nextInt(21)).south(-10 + randomsource.nextInt(21));
-                                        BlockState iblockdata = world.getBlockState(blockposition1);
+                                        // Gale start - MultiPaper - don't load chunks to spawn phantoms
+                                        BlockState iblockdata;
+                                        if (world.galeConfig().smallOptimizations.loadChunks.toSpawnPhantoms) {
+                                            iblockdata = world.getBlockState(blockposition1);
+                                        } else {
+                                            iblockdata = world.getBlockStateIfLoaded(blockposition1);
+                                            if (iblockdata == null) continue;
+                                        }
+                                        // Gale end - MultiPaper - don't load chunks to spawn phantoms
                                         FluidState fluid = world.getFluidState(blockposition1);
 
                                         if (NaturalSpawner.isValidEmptySpawnBlock(world, blockposition1, iblockdata, fluid, EntityType.PHANTOM)) {
diff --git a/src/main/java/org/galemc/gale/configuration/GaleWorldConfiguration.java b/src/main/java/org/galemc/gale/configuration/GaleWorldConfiguration.java
index 63ff60f9d5d9a8a5ea36754dd394696b54082d93..0c1a45ae1e4633a873a37ad4382b3a365331b0fa 100644
--- a/src/main/java/org/galemc/gale/configuration/GaleWorldConfiguration.java
+++ b/src/main/java/org/galemc/gale/configuration/GaleWorldConfiguration.java
@@ -92,6 +92,11 @@ public class GaleWorldConfiguration extends ConfigurationPart {
 
         }
 
+        public LoadChunks loadChunks;
+        public class LoadChunks extends ConfigurationPart {
+            public boolean toSpawnPhantoms = false; // Gale - MultiPaper - don't load chunks to spawn phantoms
+        }
+
     }
 
     public GameplayMechanics gameplayMechanics;
